bittide/src/Bittide/Df.hs

```diff
@@ -0,0 +1,132 @@
+stream stops writing characters.
+-}
+asciiDebugMux ::
+  forall timeout dom n m.
+  ( HiddenClockResetEnable dom
+  , KnownNat n
+  , KnownNat m
+  , 1 <= n
+  , 1 <= m
+  , 1 <= timeout
+  ) =>
+  -- | Timeout after /n/ cycles. Recommended setting is 1024 cycles for CPUs.
+  SNat timeout ->
+  -- | Labels to insert before each message
+  Vec n (Vec m Byte) ->
+  Circuit
+    (Vec n (Df dom Byte))
+    (Df dom Byte)
+asciiDebugMux sTimeout labels0 = circuit $ \dfs -> do
+  packets <- timeouts <| bytesToPacketStreams -< dfs
+  muxedDfs <- packetArbiterC Skip -< packets
+  -- Safety: though the function below is marked unsafe, it is safe to use here,
+  -- because we don't expect to generate any /abort/ transfers and the packet
+  -- boundaries are used internally in 'asciiDebugMux' only.
+  unsafeBytesFromPacketStream (labels1 !!) -< muxedDfs
+ where
+  bytesToPacketStreams = bundleC $ bytesToPacketStream (== newline) <$> indicesI @n
+  timeouts = bundleC $ timeout sTimeout <$> breakupM2Ss
+  labels1 = (\l -> openBracket :> l ++ closeBracket :> space :> Nil) <$> labels0
+
+  breakupM2Ss :: Vec n (PacketStreamM2S 1 (Index n))
+  breakupM2Ss =
+    indicesI <&> \ix ->
+      PacketStreamM2S
+        { _data = newline :> Nil
+        , _meta = ix
+        , _last = Just 1
+        , _abort = False
+        }
+
+  -- Characters needed to format the an output to "[${label}] ${message}\n"
+  (newline, openBracket, closeBracket, space) =
+    ( fromIntegral $ ord '\n'
+    , fromIntegral $ ord '['
+    , fromIntegral $ ord ']'
+    , fromIntegral $ ord ' '
+    )
+
+{- | Convert a 'Df' stream of bytes into a packet stream of bytes. The supplied
+predicate is used to determine the end of a packet. The meta information for
+the packet stream is constant and is supplied as an argument.
+-}
+bytesToPacketStream ::
+  -- | Predicate to determine packet delimiter
+  (Byte -> Bool) ->
+  -- | Meta information for the packet stream
+  meta ->
+  Circuit (Df dom Byte) (PacketStream dom 1 meta)
+bytesToPacketStream predicate meta = Circuit (unbundle . fmap go . bundle)
+ where
+  go (m2s, s2m) = (coerce s2m, fmap toM2S m2s)
+
+  toM2S dat =
+    PacketStreamM2S
+      { _data = dat :> Nil
+      , _meta = meta
+      , _last = if predicate dat then Just 1 else Nothing
+      , _abort = False
+      }
+
+{- | Convert a packet stream of bytes into a 'Df' stream of bytes. Is marked
+unsafe, because it drops transfers marked /abort/ and loses information about
+packet boundaries.
+-}
+unsafeBytesFromPacketStream ::
+  ( HiddenClockResetEnable dom
+  , KnownNat n
+  , 1 <= n
+  ) =>
+  (meta -> Vec n Byte) ->
+  Circuit (PacketStream dom 1 meta) (Df dom Byte)
+unsafeBytesFromPacketStream f =
+  packetizerC (const ()) f |> unsafePacketStreamToDf
+ where
+  unsafePacketStreamToDf = Circuit (unbundle . fmap go . bundle)
+   where
+    go (m2s, ack)
+      | Just PacketStreamM2S{_abort = True} <- m2s = (coerce True, Nothing)
```

Comment:
The `PacketStream` protocol does not require the `_abort` signal to stay asserted during the remainder of the packet, if you drop when `_abort` is set, you might just drop a single transfer. I don't think this is what you want.

Given that it is already unsafe, I think it would be better for this component to ignore the `_abort` signal and we should add a component that:
* When `_abort` is set, it truncates the packet by setting `_last` to `Just 0`.
* Consumes all following transactions up to the actual end of the packet without forwarding them.

Reply:
See https://github.com/clash-lang/clash-protocols/pull/159 :)

Reply:
Thanks, I'll add it here. I'll postpone upstreaming `timeout` a bit, because I would need to figure out how to write tests for it..